### PR TITLE
Ensure automated PRs are created, not duplicated

### PR DESCRIPTION
**Pull request by samuelallan72**
Previously there was the issue where a new PR to target repositories was created each time the terraform apply workflow runs, if there were outstanding changes to files.
This caused a lot of noise.

By using a consistent branch name and some extra logic,
we can ensure existing PRs are updated,
rather than creating new ones.

Fixes: #63

### CODE REVIEW

**Comment by samuelallan72** [comment]
The method of simply hardcoding the branch for the PR does work, but the terraform resource to create the PR fails:

> Error: POST https://api.github.com/repos/canonical/charm-nginx/pulls: 422 Validation Failed [{Resource:PullRequest Field: Code:custom Message:A pull request already exists for canonical:automation/update-managed-files.}]

**Comment by samuelallan72** [comment]
It succeeds if the target PR is closed.

**Review by Pjack — DISMISSED** [review]

**Comment by chanchiwai-ray** [comment]
will it update the PR or just don't create new PR?

**Comment by samuelallan72** [comment]
> will it update the PR or just don't create new PR?

@chanchiwai-ray if there is no existing PR, it will create it.  If there is an existing PR, it will update it.

**Review by chanchiwai-ray — APPROVED** [review]

File: terraform-plans/modules/GitHub/templates/main.tf
Comment (rgildein) [inline terraform-plans/modules/GitHub/templates/main.tf:64]: You did not remove random_string (line 15-20).

**Review by rgildein — DISMISSED** [review]
What if we open two PR here and one is approved -> new PR for targeted repos is created and before such PR (in targeted repo) is approved and merge, another PR in this repo is merged.
I think that this will cause silently not doing any new PR and missing those changes. 
Can we instead of random branch name use some id/hash/... from git commit? e.g. `automation/update-managed-file/dd1bc07`

File: terraform-plans/modules/GitHub/templates/main.tf
Comment (samuelallan72) [reply terraform-plans/modules/GitHub/templates/main.tf:64]: Thanks, updated.

**Comment by samuelallan72** [comment]
> What if we open two PR here and one is approved -> new PR for targeted repos is created and before such PR (in targeted repo) is approved and merge, another PR in this repo is merged. I think that this will cause silently not doing any new PR and missing those changes.

Sorry I can't follow what you're saying here.

If there's an already open PR, that PR will be updated with new commits.  So we won't miss the changes; we'll see them when reviewing the target repo automated PR.

**Review by chanchiwai-ray — APPROVED** [review]

**Review by Pjack — APPROVED** [review]

**Comment by Pjack** [comment]
> > What if we open two PR here and one is approved -> new PR for targeted repos is created and before such PR (in targeted repo) is approved and merge, another PR in this repo is merged. I think that this will cause silently not doing any new PR and missing those changes.
> 
> Sorry I can't follow what you're saying here.
> 
> If there's an already open PR in the target repo, that PR will be updated with new commits. So we won't miss the changes; we'll see them when reviewing the target repo automated PR.

@rgildein 
I don't fully understand your wording, but I assume your main concern is that some changes might be merged without review. This could happen if the changes are made after the automated PR has been approved but not merged yet.
We've enabled the 'Dismiss stale pull request approvals when new commits are pushed' setting in every repository, so this issue should no longer happen.

https://github.com/canonical/solutions-engineering-automation/blob/main/terraform-plans/modules/GitHub/settings/main.tf#L82

**Comment by rgildein** [comment]
> > What if we open two PR here and one is approved -> new PR for targeted repos is created and before such PR (in targeted repo) is approved and merge, another PR in this repo is merged. I think that this will cause silently not doing any new PR and missing those changes.
> 
> Sorry I can't follow what you're saying here.
> 
> If there's an already open PR in the target repo, that PR will be updated with new commits. So we won't miss the changes; we'll see them when reviewing the target repo automated PR.

I see, so one PR could have multiple changes from multiple automated commits? If yes, is this what we want? Since what I know, we are doing squash and merge, so multiple changes will be done by single PR <==> single commit. Maybe it's fine, since these changes are done by automation. Not Sure.

**Comment by samuelallan72** [comment]
@rgildein 

> I see, so one PR could have multiple changes from multiple automated commits? If yes, is this what we want?

Yep that's correct. :)  It should be fine, because it's automated - the commit messages and description of the PR doesn't tell us much already; the real change history is in this automation repo.

**Comment by rgildein** [comment]
> Yep that's correct. :) It should be fine, because it's automated - the commit messages and description of the PR doesn't tell us much already; the real change history is in this automation repo.

I agree :+1:, originally I thought that if PR already exists, new changes will not be applied. Thanks!

**Pull request merged**
